### PR TITLE
Issue #936: Refactor Unit Test

### PR DESCRIPTION
**Pull request by mayankag32**
Results from unittest.exe - 
`Execution finished in 32.298 seconds.
Failed: 0 Passed: 56 Skipped: 105 Percentage: 100.00%`

[ Please outline your change here. Be sure to check the checkboxes below. ]

## Code change checklist

- [x] I have ensured that all methods and functions are **fully documented** using doxygen style comments.
- [x] My code follows the [coding style guide](https://dpp.dev/coding-standards.html).
- [x] I tested that my change works before raising the PR.
- [x] I have ensured that I did not break any existing API calls.
- [x] I have not built my pull request using AI, a static analysis tool or similar without any human oversight.

### CODE REVIEW

**Comment by braindigitalis** [comment]
105 tests were skipped, does this still work with online tests?

File: src/unittest/test.h [hunk 1]
```diff
@@ -31,8 +31,6 @@ _Pragma("warning( disable : 5105 )"); // 4251 warns when we export classes or st
 
 #ifdef _WIN32
 #define SHARED_OBJECT "dpp.dll"
-#elif __APPLE__
```
Comment (braindigitalis) [inline src/unittest/test.h:34]: why is this removed?

File: src/unittest/test.h [hunk 2]
```diff
@@ -310,9 +307,9 @@ inline constexpr bool coro = false;
 		}); \
 	}
 
-/**
- * @brief Perform a test of a REST base API call with one parameter
- */
+ /**
```
Comment (braindigitalis) [inline src/unittest/test.h:310]: why are spaces added here

File: src/unittest/test.h [hunk 3]
```diff
@@ -150,7 +148,6 @@ DPP_TEST(FORUM_CHANNEL_GET, "retrieve the created forum channel", tf_online);
 DPP_TEST(FORUM_CHANNEL_DELETE, "delete the created forum channel", tf_online);
 DPP_TEST(ERRORS, "Human readable error translation", tf_offline);
 
-DPP_TEST(GUILD_EDIT, "cluster::guild_edit", tf_online);
```
Comment (braindigitalis) [inline src/unittest/test.h:153]: was this test moved elsewhere?

File: src/unittest/test.h [hunk 4]
```diff
@@ -332,9 +329,9 @@ inline constexpr bool coro = false;
 		}); \
 	}
 
-/**
- * @brief Perform a test of a REST base API call with one parameter that returns a list
- */
+  /**
```
Comment (braindigitalis) [inline src/unittest/test.h:332]: weird spacing again

File: src/unittest/test.h [hunk 5]
```diff
@@ -416,73 +413,73 @@ inline constexpr bool coro = false;
...
- * @param test The test to set the status of
- * @param success If set to true, sets success to true, if set to false and called
- * once, sets executed to true, if called twice, also sets success to false.
- * This means that before you run the test you should call this function once
- * with success set to false, then if/wen the test completes call it again with true.
- * If the test fails, call it a second time with false, or not at all.
- */
-void set_test(test_t &test, bool success = false);
+	  /**
+	   * @brief Sets a test's status (legacy)
```
Comment (braindigitalis) [inline src/unittest/test.h:417]: spacing weirdness

**Review by braindigitalis — CHANGES_REQUESTED** [review]
please check requested changes :)

**Comment by mayankag32** [comment]
unit test with token - 
```
UNIT TEST SUMMARY
==================
dpp::snowflake class                                         PASS
dpp::json_interface class                                    PASS
Instantiate DPP cluster                                      FAIL
cluster::start method                                        NOT EXECUTED
Connection to client websocket                               FAIL
Creation of application command                              NOT EXECUTED
Deletion of application command                              NOT EXECUTED
Log events                                                   NOT EXECUTED
Creation of a channel message                                NOT EXECUTED
Editing a channel message                                    NOT EXECUTED
Message edit event                                           NOT EXECUTED
Deletion of a channel message                                NOT EXECUTED
Receipt of a created message                                 NOT EXECUTED
Message attachment send and check                            NOT EXECUTED
Test guild cache                                             NOT EXECUTED
Test user cache                                              NOT EXECUTED
React to a message                                           NOT EXECUTED
Reaction event                                               NOT EXECUTED
Receive guild create event                                   FAIL
Get messages                                                 NOT EXECUTED
crossplatform_strptime()                                     NOT EXECUTED
utility::iconhash                                            PASS
cluster::current_user_get()                                  NOT EXECUTED
cluster::guild_get()                                         NOT EXECUTED
cluster::channel_get()                                       NOT EXECUTED
cluster::channels_get()                                      NOT EXECUTED
cluster::roles_get()                                         NOT EXECUTED
cluster::guild_get_invites()                                 NOT EXECUTED
cluster::guild_get_bans()                                    NOT EXECUTED
cluster::channel_pins_get()                                  NOT EXECUTED
cluster::guild_events_get()                                  NOT EXECUTED
cluster::guild_event_get()                                   NOT EXECUTED
message_create_t::send()                                     NOT EXECUTED
cluster::guild_event_users_get()                             NOT EXECUTED
start timer                                                  NOT EXECUTED
stop timer                                                   NOT EXECUTED
one-shot timer                                               NOT EXECUTED
timed listener                                               NOT EXECUTED
Presence intent                                              FAIL
Instantiate a cache                                          PASS
message_collector                                            FAIL
managed::get_creation_date()                                 PASS
utility::read_file()                                         PASS
ts_to_string()                                               PASS
ts_not_null()                                                PASS
command_option_choice::fill_from_json: double                PASS
command_option_choice::fill_from_json: int64_t               PASS
command_option_choice::fill_from_json: bool                  PASS
command_option_choice::fill_from_json: snowflake             PASS
command_option_choice::fill_from_json: string                PASS
https_client::get_host_info()                                PASS
https_client HTTPS request                                   FAIL
https_client HTTP request                                    PASS
run_once<T>                                                  PASS
webhook construct from URL                                   PASS
Markdown escaping (ignore code block contents)               PASS
Markdown escaping (escape code block contents)               PASS
URL encoding                                                 PASS
Base 64 encoding                                             PASS
sync<T>()                                                    FAIL
manged object comparison                                     PASS
find_channel()                                               NOT EXECUTED
channel type flags                                           NOT EXECUTED
create a forum channel                                       NOT EXECUTED
retrieve the created forum channel                           NOT EXECUTED
delete the created forum channel                             NOT EXECUTED
Human readable error translation                             PASS
cluster::guild_ban_add ban three deleted discord accounts    NOT EXECUTED
cluster::guild_get_ban getting one of the banned accounts    NOT EXECUTED
cluster::guild_get_bans get bans using the after-parameter   NOT EXECUTED
cluster::guild_ban_delete unban the banned discord accounts  NOT EXECUTED
cluster::thread_create                                       NOT EXECUTED
cluster::on_thread_create event                              NOT EXECUTED
cluster::channel_delete with thread                          NOT EXECUTED
cluster::on_thread_delete event                              NOT EXECUTED
cluster::thread_edit                                         NOT EXECUTED
cluster::on_thread_update event                              NOT EXECUTED
cluster::threads_get_active                                  NOT EXECUTED
creating a voice channel                                     NOT EXECUTED
editing the created voice channel                            NOT EXECUTED
deleting the created voice channel                           NOT EXECUTED
permission                                                   PASS
event class                                                  PASS
cluster::user_get                                            NOT EXECUTED
cluster::user_get flag parsing                               NOT EXECUTED
cluster::guild_get_member                                    NOT EXECUTED
user::get_mention                                            PASS
user::get_url                                                PASS
user::format_username                                        PASS
user::get_creation_time                                      PASS
user::get_avatar_url                                         PASS
channel::set_type                                            PASS
channel::get_mention                                         PASS
channel::get_url                                             PASS
message::get_url                                             PASS
utility::guild_navigation                                    PASS
utility::iconhash                                            PASS
utility::make_url_parameters                                 PASS
utility::markdown_escape                                     PASS
utility::tokenize                                            PASS
utility::url_encode                                          PASS
utility::slashcommand_mention                                PASS
utility::channel_mention                                     PASS
utility::user_mention                                        PASS
utility::role_mention                                        PASS
utility::emoji_mention                                       PASS
utility::user_url                                            PASS
utility::message_url                                         PASS
utility::channel_url                                         PASS
utility::thread_url                                          PASS
utility::avatar_size                                         PASS
utility::cdn_endpoint_url_hash                               PASS
sticker::get_url aka utility::cdn_endpoint_url_sticker       PASS
emoji::get_url                                               PASS
role::operator<                                              PASS
cluster::role_create                                         NOT EXECUTED
cluster::role_edit                                           NOT EXECUTED
cluster::role_delete                                         NOT EXECUTED
JSON parse error for post_rest                               PASS
cluster::user_get_cached_sync() with present member          FAIL
cluster::user_get_cached_sync() with not present member      PASS
interaction_create_t::get_parameter() with subcommands       PASS
interaction_create_t::get_parameter() without subcommands    PASS
cluster::automod_rule_create                                 NOT EXECUTED
cluster::automod_rule_get                                    NOT EXECUTED
cluster::automod_rules_get                                   NOT EXECUTED
cluster::automod_rule_delete                                 NOT EXECUTED
using the cluster::request method to fetch an image          NOT EXECUTED
cluster::guild_emoji_create                                  NOT EXECUTED
cluster::guild_emoji_get                                     NOT EXECUTED
cluster::guild_emoji_delete                                  NOT EXECUTED
cluster::on_invite_create                                    NOT EXECUTED
cluster::on_invite_delete                                    NOT EXECUTED
cluster::channel_invite_create                               NOT EXECUTED
cluster::invite_get                                          NOT EXECUTED
cluster::invite_delete                                       NOT EXECUTED
multiheader cookie test                                      SKIPPED
Connect to voice channel                                     SKIPPED
Send audio to voice channel                                  SKIPPED
Pinning a channel message                                    SKIPPED
Unpinning a channel message                                  SKIPPED
cluster::thread_member_add                                   SKIPPED
cluster::thread_member_get                                   SKIPPED
cluster::thread_members_get                                  SKIPPED
cluster::thread_member_remove                                SKIPPED
cluster::on_thread_members_update event with member addition SKIPPED
cluster::on_thread_members_update event with member removal  SKIPPED
cluster::thread_create_with_message                          SKIPPED
message manipulation in thread                               SKIPPED
cluster::on_message_create in thread                         SKIPPED
cluster::on_message_edit in thread                           SKIPPED
cluster::on_message_delete in thread                         SKIPPED
cluster::on_reaction_add in thread                           SKIPPED
cluster::on_reaction_remove in thread                        SKIPPED
coro: offline job                                            SKIPPED
coro: offline coroutine                                      SKIPPED
coro: offline task                                           SKIPPED
coro: offline async                                          SKIPPED
coro: online event handler                                   SKIPPED
coro: online api calls                                       SKIPPED
coro: online mumbo jumbo in event handler                    SKIPPED

Execution finished in 3.441 seconds.
Failed: 78 Passed: 58 Skipped: 25 Percentage: 42.65%
```

**Comment by Mishura4** [comment]
Thanks for contributing!

Make sure you run the full suite of tests including online tests. You'll need to set a couple environment variables, see https://dpp.dev/unit-tests.html

Also pass "full" as the first command argument for some extra tests

**Comment by mayankag32** [comment]
> Thanks for contributing!
> 
> Make sure you run the full suite of tests including online tests. You'll need to set a couple environment variables, see https://dpp.dev/unit-tests.html
> 
> Also pass "full" as the first command argument for some extra tests

Thanks for the suggestion, I was looking for environment variables details.

**Comment by braindigitalis** [comment]
fails on mac os:

https://github.com/brainboxdotcc/DPP/actions/runs/7863356947/job/21453913274?pr=1080

File: src/unittest/test.h (hunk 1 above)
Comment (Jaskowicz1) [reply src/unittest/test.h:34]: Please revert this change. This was specifically added to fix unit tests breaking on macs.

This now breaks all our CIs for mac, along with any future unit test (if approved).

**Review by Jaskowicz1 — CHANGES_REQUESTED** [review]
Thanks for contributing! Your PR seems to be a near copy-paste of #948 with some minor modifications. I'm also assuming you based this PR off that PR as you're attempting to remove Mac support.

Please correct the Mac support issue, along with the fails shown in your results!

**Comment by Jaskowicz1** [comment]
Hi there, are you currently working on readding mac support and fixing the other issues? You seemed to fix the space indentation but left everything else.

**Comment by Jaskowicz1** [comment]
We've closed this PR due to no activity. The PR also seems to be a direct copy of an old PR that was raised, with some additions. This also breaks our OSX support.
